dev/compose/aap.yaml

```diff
@@ -0,0 +1,329 @@
+  PULP_ANSIBLE_BASE_JWT_VALIDATE_CERT: 'false'
+  PULP_ANSIBLE_BASE_JWT_KEY: 'https://localhost'
+  PULP_GALAXY_FEATURE_FLAGS__external_authentication: 'true'
+
+  # disable user/group modifications
+  PULP_ALLOW_LOCAL_RESOURCE_MANAGEMENT: 'false'
+
+  # role content workaround ..
+  PULP_ANSIBLE_BASE_ROLES_REQUIRE_VIEW: 'false'
+
+  # DEV EDITABLE STUFF
+  LOCK_REQUIREMENTS: 0
+  DEV_SOURCE_PATH: "galaxy_ng"
+  # To enable editable installs of local checkouts change the variable above keeping the ordering as follows:
+  # DEV_SOURCE_PATH: "dynaconf:pulpcore:galaxy_importer:pulp_ansible:pulp_container:galaxy_ng:django-ansible-base"
+
+
+services:
+  base_img:
+    build:
+      context: ..
+      dockerfile: Dockerfile
+    image: "localhost/galaxy_ng/galaxy_ng:base"
+
+  redis:
+    image: "redis:5"
+
+  postgres:
+    image: "postgres:13"
+    environment:
+      <<: *common-env
+    healthcheck:
+      test: ["CMD", "pg_isready", "-U", "galaxy_ng"]
+      interval: 10s
+      retries: 5
+
+  helper:  # should this be moved to a custom Dockerfile?
+    image: quay.io/centos/centos:stream9
+    environment:
+      <<: *common-env
+    depends_on:
+      - postgres
+    volumes:
+      - "etc_pulp_certs:/etc/pulp/certs"
+      - "var_lib_pulp:/var/lib/pulp"
+      - "../../../:/src"
+    command: |
+      bash -c "
+        echo 'Moving required files to the required places.';
+
+        # Workarounds - Pulp SigningService can't access envvars;
+        /src/galaxy_ng/dev/compose/signing/setup_gpg_workarounds.sh;
+        cat /etc/pulp/certs/GNUPGHOME.workaround.txt;
+        cat /etc/pulp/certs/HOME.workaround.txt;
+
+        # Keys;
+        cp /src/galaxy_ng/dev/compose/database/database_fields.symmetric.key /etc/pulp/certs/database_fields.symmetric.key;
+        cp /src/galaxy_ng/dev/compose/signing/signing-secret.key /etc/pulp/certs/signing-secret.key
+        cp /src/galaxy_ng/dev/compose/signing/signing-secret.key.password.txt /etc/pulp/certs/signing-secret.key.password.txt
+        cp /src/galaxy_ng/dev/compose/signing/signing-secret.key /etc/pulp/certs/signing-public.key
+        find /etc/pulp/certs;
+
+        # Scripts;
+        cp /src/galaxy_ng/dev/compose/signing/collection_sign.sh /var/lib/pulp/scripts/collection_sign.sh;
+        cp /src/galaxy_ng/dev/compose/signing/container_sign.sh /var/lib/pulp/scripts/container_sign.sh;
+        chmod +x /var/lib/pulp/scripts/*_sign.sh;
+        find /var/lib/pulp/scripts;
+
+        echo 'DONE!';
+      "
+
+  migrations:
+    image: "localhost/galaxy_ng/galaxy_ng:base"
+    depends_on:
+      - base_img
+      - postgres
+      - helper
+    volumes:
+      - "etc_pulp_certs:/etc/pulp/certs"
+      - "var_lib_pulp:/var/lib/pulp"
+      - "../../../:/src"
+    environment:
+      <<: *common-env
+    user: root
+    command: |
+      bash -c "
```

Comment:
@cutwater We are trying to Over-Abstraction and multiple layers of indirection.

The initial version of this PR had scripts encoded as `base64` inlined in the yaml file, then on those cases we agreed to move to external script.

I think the advantage is that reading the YAML would be enough to understand at least the higher level step-by-step that is being executed to prepare the environment.

A scenario where I see this being helpful is when hacking/debugging, and I want to hook some other calls or experiment a different parameter for example on a management command.

I can just copy/paste the yaml file as `aap_hacking.yaml` and do my changes with little changes to sparse bash scripts that are nested in a way we can´t easily tell or explain.

We had this experience before on the old `./compose` environment, on `pplugin_template` and again on `oci-env`

The goal is that reading the `aap.yaml` only, anyone can understand the step-by-step and manipulate in the way it is needed to develop and debug.

I think we can encapsulate some atomic actions as scripts, but reading the yaml needs to make it obvious what is happening.

Also, I see advantage on moving some things to a dev Dockerfile, but we will discuss it further if needed, we are doing baby-step improvements to avoid over-engineering.

Reply:
I think the rule of thumb here would be to stick to the SOLID `S` principle.

An external script will do One Single thing only, never write scripts that performs nested calls to other scripts.

Multiple scripts must be chained explicitly on the compose yaml and with meaningful name that describes what the script is doing.


A refactor that would work but still keep the visibility would be:
```bash
bash -c "
   /src/galaxy_ng/dev/compose/bin/wait /etc/pulp/certs/database_fields.symmetric.key
   /src/galaxy_ng/dev/compose/bin/wait /etc/pulp/certs/signing-public.key
   /src/galaxy_ng/dev/compose/bin/run_migrations
"
```

What we want to avoid is:

```
command: magic_entrypoint.sh
```